src/main/java/io/cdap/plugin/hubspot/common/HubspotHelper.java

```
@@ -50,73 +57,104 @@ public class HubspotHelper {
+    CloseableHttpResponse response = executeRequestWithRetries(getRequest(config, offset));
     HttpEntity entity = response.getEntity();
     String result;
     if (entity != null) {
       result = EntityUtils.toString(entity);
-      return parseJson(sourceHubspotConfig, result);
+      return parseJson(config, result);
     }
     return null;
   }
 
-  private CloseableHttpResponse downloadPage(SourceHubspotConfig sourceHubspotConfig,
-                                             HttpGet request) throws IOException {
+  /** Executes the given request until it's successful
+   * or the maximum attempts number is exceeded (then {@link IOException} is thrown). */
+  public static CloseableHttpResponse executeRequestWithRetries(HttpRequestBase request) throws IOException {
+    return executeRequestWithRetries(request, MAX_RETRIES_DEFAULT);
+  }
+
+  /** Executes the given request until it's successful
+   * or maximum retries attempts is exceeded (then {@link IOException} is thrown). */
+  public static CloseableHttpResponse executeRequestWithRetries(HttpRequestBase request, int maxRetries)
+          throws IOException {
     HttpClientBuilder httpClientBuilder = HttpClientBuilder.create();
     CloseableHttpClient client = httpClientBuilder.build();
 
     int count = 0;
-    while (true) {
+    StatusLine statusLine = null;
+    while (count <= maxRetries) {
+      ++count;
       CloseableHttpResponse response = client.execute(request);
-      StatusLine statusLine = response.getStatusLine();
-      if (statusLine.getStatusCode() >= 300) {
-        if (500 > statusLine.getStatusCode() &&
-          statusLine.getStatusCode() >= 400) {
-          throw new IOException(statusLine.getReasonPhrase());
-        }
-        if (++count == MAX_TRIES) {
-          throw new IOException(statusLine.getReasonPhrase());
-        }
+      statusLine = response.getStatusLine();
+      if (200 <= statusLine.getStatusCode() && statusLine.getStatusCode() < 300) {
```

Can we throw an exception immediately if the response is 400 or 403? Those status codes typically do not indicate a retryable exception.

---

Done